File: packages/cli/src/ignore-files.ts
/**
 * Vendored docs ignore-file management.
 *
 * `.ask/docs/` contains third-party documentation downloaded by ASK. It
 * should not be treated as project source: lint/format/code-review tools
 * should skip it, but AI agents must still be able to read it as context.
 *
 * Strategy:
 *   A. Write self-contained nested config files inside `.ask/docs/` for
 *      tools that support hierarchical config resolution (Git, ESLint flat
 *      config, Biome, markdownlint-cli2).
 *   B. The "intent notice" for AI review tools lives in `agents.ts`
 *      (extended `generateAgentsMd`).
 *   C. Patch root files only for tools that do NOT support nested config
 *      (Prettier, SonarQube, legacy markdownlint-cli).
 */

import fs from 'node:fs'
import path from 'node:path'
import { consola } from 'consola'
import { loadConfig } from './config.js'
import { inject, remove as removeMarker, wrap } from './markers.js'
import { getDocsDir } from './storage.js'

/**
 * Local configuration files written inside `.ask/docs/` so that
 * lint/format/review tools with nested-config support automatically skip
 * the directory. Each file has a `name` (relative to `.ask/docs/`) and a
 * `content` template.
 */
const NESTED_CONFIGS: Array<{ name: string, content: string }> = [
  {
    name: '.gitattributes',
    content: [
      '# Managed by ASK — marks vendored docs for GitHub Linguist.',
      '* linguist-vendored=true',
      '* linguist-generated=true',
      '',
    ].join('\n'),
  },
  {
    name: 'eslint.config.mjs',
    content: [
      '// Managed by ASK — vendored docs, excluded from ESLint.',
      'export default [',
      '  { ignores: [\'**/*\'] },',
      ']',
      '',
    ].join('\n'),
  },
  {
    name: 'biome.json',
    content: `${JSON.stringify(
      {
        $schema: 'https://biomejs.dev/schemas/2.0.0/schema.json',
        files: { ignore: ['**/*'] },
      },
      null,
      2,
    )}\n`,
  },
  {
    name: '.markdownlint-cli2.jsonc',
    content: `${JSON.stringify(
      { ignores: ['**/*'] },
      null,
      2,
    )}\n`,
  },
]

interface RootPatch {
  /** File name relative to project root. */
  file: string
  /** Marker syntax to use when injecting/removing. */
  syntax: 'html' | 'hash'
  /** Payload to wrap between markers. */
  payload: string
  /** Optional warning emitted when the file is detected. */
  warn?: string
}

const ROOT_PATCHES: RootPatch[] = [
  {
    file: '.prettierignore',
    syntax: 'hash',
    payload: '# Vendored docs — managed by ASK\n.ask/docs/',
  },
  {
    file: 'sonar-project.properties',
    syntax: 'hash',
    payload: '# Vendored docs — managed by ASK\nsonar.exclusions=.ask/docs/**',
  },
  {
    file: '.markdownlintignore',
    syntax: 'hash',
    payload: '# Vendored docs — managed by ASK\n.ask/docs/',
    warn: 'Legacy .markdownlintignore detected. Consider migrating to markdownlint-cli2, which supports nested config inside .ask/docs/ automatically.',
  },
]

/**
 * Category A: write nested config files inside `.ask/docs/`.
 *
 * The directory is created if it does not exist, so callers may invoke this
 * before any docs have been saved. Existing files are only rewritten if
 * their contents differ, to keep filesystem mtimes stable and logs terse.
 *
 * Returns the list of files that were created or updated (relative to the
 * project root) so callers can log a summary.
 */
export function writeNestedConfigs(projectDir: string): string[] {
  const docsDir = getDocsDir(projectDir)
  fs.mkdirSync(docsDir, { recursive: true })

  const written: string[] = []
  for (const { name, content } of NESTED_CONFIGS) {
    const target = path.join(docsDir, name)
    const existing = fs.existsSync(target) ? fs.readFileSync(target, 'utf-8') : null
    if (existing !== content) {
      fs.writeFileSync(target, content, 'utf-8')
      written.push(path.relative(projectDir, target))
    }
  }
  return written
}

/**
 * Remove all nested config files written by {@link writeNestedConfigs}.
 *
 * Only files with the exact names we manage are deleted. Other files
 * inside `.ask/docs/` are left alone (including downloaded docs). Returns
 * the list of removed file paths relative to the project root.
 */
export function removeNestedConfigs(projectDir: string): string[] {
  const docsDir = getDocsDir(projectDir)
  if (!fs.existsSync(docsDir))
    return []
  const removed: string[] = []
  for (const { name } of NESTED_CONFIGS) {
    const target = path.join(docsDir, name)
    if (fs.existsSync(target)) {
      fs.rmSync(target)
      removed.push(path.relative(projectDir, target))
    }
  }
  return removed
}

/**
 * Category C: patch root files that do not support nested ignore resolution.
 *
 * Only files that already exist are patched — ASK never creates a root
 * ignore file from scratch, because doing so could imply a tool the user
 * does not actually use.
 */
export function patchRootIgnores(projectDir: string): string[] {
  const updated: string[] = []
  for (const patch of ROOT_PATCHES) {
    const target = path.join(projectDir, patch.file)
    if (!fs.existsSync(target))
      continue

    const existing = fs.readFileSync(target, 'utf-8')
    const block = wrap(patch.payload, patch.syntax)
    const next = inject(existing, block, patch.syntax)
    if (next !== existing) {
      fs.writeFileSync(target, next, 'utf-8')
      updated.push(patch.file)
    }
    if (patch.warn) {
      consola.warn(patch.warn)
    }
  }
  return updated
}

/**
 * Remove ASK-managed marker blocks from root files patched by
 * {@link patchRootIgnores}. Files themselves are never deleted.
 */
export function unpatchRootIgnores(projectDir: string): string[] {
  const updated: string[] = []
  for (const patch of ROOT_PATCHES) {
    const target = path.join(projectDir, patch.file)
    if (!fs.existsSync(target))
      continue
    const existing = fs.readFileSync(target, 'utf-8')
    const next = removeMarker(existing, patch.syntax)
    if (next !== existing) {
      fs.writeFileSync(target, next, 'utf-8')
      updated.push(patch.file)
    }
  }
  return updated
}

/**
 * Top-level orchestrator called from the add/sync/remove commands.
 *
 * - `install` mode: create nested configs and patch detected root files.
 * - `remove`  mode: delete nested configs and strip root marker blocks.
 *
 * Respects `manageIgnores` in `.ask/config.json` (default: true). When the
 * flag is explicitly set to false, the function is a no-op.
 */
export function manageIgnoreFiles(
  projectDir: string,
  mode: 'install' | 'remove',
): void {
  let manage = true
  try {
    const config = loadConfig(projectDir)
    if (config.manageIgnores === false)
      manage = false
  }
  catch {
    // No config yet — default to managing. This matches the behaviour of
    // a fresh project where `ask docs add` writes the config on the fly.
  }

  if (!manage) {
    consola.info('Skipping ignore-file management (manageIgnores: false).')
    return
  }

  if (mode === 'install') {
    const nested = writeNestedConfigs(projectDir)
    const root = patchRootIgnores(projectDir)
    if (nested.length > 0)
      consola.info(`Nested configs written: ${nested.join(', ')}`)
    if (root.length > 0)
      consola.info(`Root ignore files patched: ${root.join(', ')}`)
  }
  else {
    const nested = removeNestedConfigs(projectDir)
    const root = unpatchRootIgnores(projectDir)
    if (nested.length > 0)
      consola.info(`Nested configs removed: ${nested.join(', ')}`)
    if (root.length > 0)
      consola.info(`Root ignore marker blocks removed: ${root.join(', ')}`)
  }
}

File: crates/ask/src/ignore_files.rs
//! Vendored-docs ignore-file management. Rust port of `ignore-files.ts`.
//!
//! `.ask/docs/` holds third-party docs — lint/format/review tools should skip it
//! while agents still read it. Strategy: (A) write self-contained nested configs
//! inside `.ask/docs/` for tools with hierarchical config, (C) patch root files
//! for tools without nested config (Prettier, SonarQube, legacy markdownlint).

use std::path::Path;

use crate::io::read_ask_json;
use crate::markers::{inject, remove as remove_marker, wrap, MarkerSyntax};
use crate::storage::get_docs_dir;

/// install (create/patch) vs remove (delete/strip).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum IgnoreMode {
    Install,
    Remove,
}

struct NestedConfig {
    name: &'static str,
    content: &'static str,
}

/// Nested config files written inside `.ask/docs/` (byte-exact templates).
const NESTED_CONFIGS: &[NestedConfig] = &[
    NestedConfig {
        name: ".gitattributes",
        content: "# Managed by ASK — marks vendored docs for GitHub Linguist.\n\
                  * linguist-vendored=true\n\
                  * linguist-generated=true\n",
    },
    NestedConfig {
        name: "eslint.config.mjs",
        content: "// Managed by ASK — vendored docs, excluded from ESLint.\n\
                  export default [\n\
                  \x20 { ignores: ['**/*'] },\n\
                  ]\n",
    },
    NestedConfig {
        name: "biome.json",
        content: "{\n  \"$schema\": \"https://biomejs.dev/schemas/2.0.0/schema.json\",\n  \"files\": {\n    \"ignore\": [\n      \"**/*\"\n    ]\n  }\n}\n",
    },
    NestedConfig {
        name: ".markdownlint-cli2.jsonc",
        content: "{\n  \"ignores\": [\n    \"**/*\"\n  ]\n}\n",
    },
];

struct RootPatch {
    file: &'static str,
    payload: &'static str,
    warn: Option<&'static str>,
}

/// Root files patched only when they already exist (never created from scratch).
const ROOT_PATCHES: &[RootPatch] = &[
    RootPatch {
        file: ".prettierignore",
        payload:
            "# Vendored by ASK\n.ask/docs/\n.ask/skills/\n.ask/resolved.json\n.ask/skills-lock.json",
        warn: None,
    },
    RootPatch {
        file: "sonar-project.properties",
        payload: "# Vendored by ASK\nsonar.exclusions=.ask/docs/**,.ask/skills/**",
        warn: None,
    },
    RootPatch {
        file: ".markdownlintignore",
        payload: "# Vendored by ASK\n.ask/docs/\n.ask/skills/",
        warn: Some(
            "Legacy .markdownlintignore detected. Consider migrating to markdownlint-cli2, which \
             supports nested config inside .ask/docs/ automatically.",
        ),
    },
    RootPatch {
        file: ".gitignore",
        payload:
            "# Vendored by ASK\n.ask/docs/\n.ask/skills/\n.ask/resolved.json\n.ask/skills-lock.json",
        warn: None,
    },
];

/// Whether the user has opted into ASK via any entry point.
fn has_ask_opt_in(project_dir: &Path) -> bool {
    if matches!(read_ask_json(project_dir), Ok(Some(_))) {
        return true;
    }
    let ask = project_dir.join(".ask");
    ask.join("skills-lock.json").exists() || ask.join("skills").exists()
}

fn rel(project_dir: &Path, target: &Path) -> String {
    target
        .strip_prefix(project_dir)
        .unwrap_or(target)
        .to_string_lossy()
        .into_owned()
}

/// Category A: write nested config files inside `.ask/docs/` (only when content
/// differs). Returns project-relative paths written.
pub fn write_nested_configs(project_dir: &Path) -> anyhow::Result<Vec<String>> {
    let docs_dir = get_docs_dir(project_dir);
    std::fs::create_dir_all(&docs_dir)?;
    let mut written = Vec::new();
    for cfg in NESTED_CONFIGS {
        let target = docs_dir.join(cfg.name);
        let existing = std::fs::read_to_string(&target).ok();
        if existing.as_deref() != Some(cfg.content) {
            std::fs::write(&target, cfg.content)?;
            written.push(rel(project_dir, &target));
        }
    }
    Ok(written)
}

/// Remove the nested config files (leaving downloaded docs untouched).
pub fn remove_nested_configs(project_dir: &Path) -> anyhow::Result<Vec<String>> {
    let docs_dir = get_docs_dir(project_dir);
    if !docs_dir.exists() {
        return Ok(Vec::new());
    }
    let mut removed = Vec::new();
    for cfg in NESTED_CONFIGS {
        let target = docs_dir.join(cfg.name);
        if target.exists() {
            std::fs::remove_file(&target)?;
            removed.push(rel(project_dir, &target));
        }
    }
    Ok(removed)
}

/// Category C: inject the ASK marker block into existing root files.
pub fn patch_root_ignores(project_dir: &Path) -> anyhow::Result<Vec<String>> {
    let mut updated = Vec::new();
    for patch in ROOT_PATCHES {
        let target = project_dir.join(patch.file);
        if !target.exists() {
            continue;
        }
        let existing = std::fs::read_to_string(&target)?;
        let block = wrap(patch.payload, MarkerSyntax::Hash);
        let next = inject(&existing, &block, MarkerSyntax::Hash);
        if next != existing {
            std::fs::write(&target, &next)?;
            updated.push(patch.file.to_string());
        }
        if let Some(warn) = patch.warn {
            eprintln!("  Warning: {warn}");
        }
    }
    Ok(updated)
}

/// Strip the ASK marker block from patched root files (files never deleted).
pub fn unpatch_root_ignores(project_dir: &Path) -> anyhow::Result<Vec<String>> {
    let mut updated = Vec::new();
    for patch in ROOT_PATCHES {
        let target = project_dir.join(patch.file);
        if !target.exists() {
            continue;
        }
        let existing = std::fs::read_to_string(&target)?;
        let next = remove_marker(&existing, MarkerSyntax::Hash);
        if next != existing {
            std::fs::write(&target, &next)?;
            updated.push(patch.file.to_string());
        }
    }
    Ok(updated)
}

/// Top-level orchestrator. No-op when the user hasn't opted into ASK.
pub fn manage_ignore_files(project_dir: &Path, mode: IgnoreMode) -> anyhow::Result<()> {
    if !has_ask_opt_in(project_dir) {
        return Ok(());
    }
    match mode {
        IgnoreMode::Install => {
            let nested = write_nested_configs(project_dir)?;
            let root = patch_root_ignores(project_dir)?;
            if !nested.is_empty() {
                eprintln!("  Nested configs written: {}", nested.join(", "));
            }
            if !root.is_empty() {
                eprintln!("  Root ignore files patched: {}", root.join(", "));
            }
        }
        IgnoreMode::Remove => {
            let nested = remove_nested_configs(project_dir)?;
            let root = unpatch_root_ignores(project_dir)?;
            if !nested.is_empty() {
                eprintln!("  Nested configs removed: {}", nested.join(", "));
            }
            if !root.is_empty() {
                eprintln!("  Root ignore marker blocks removed: {}", root.join(", "));
            }
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn opt_in(project: &Path) {
        std::fs::write(crate::io::get_ask_json_path(project), r#"{"libraries":[]}"#).unwrap();
    }

    #[test]
    fn no_op_without_opt_in() {
        let dir = tempfile::tempdir().unwrap();
        // A .gitignore exists but no ask.json → untouched.
        std::fs::write(dir.path().join(".gitignore"), "node_modules\n").unwrap();
        manage_ignore_files(dir.path(), IgnoreMode::Install).unwrap();
        assert_eq!(
            std::fs::read_to_string(dir.path().join(".gitignore")).unwrap(),
            "node_modules\n"
        );
    }

    #[test]
    fn install_writes_nested_and_patches_existing_root() {
        let dir = tempfile::tempdir().unwrap();
        opt_in(dir.path());
        std::fs::write(dir.path().join(".gitignore"), "node_modules\n").unwrap();
        // .prettierignore does NOT exist → must not be created.
        manage_ignore_files(dir.path(), IgnoreMode::Install).unwrap();

        let docs = get_docs_dir(dir.path());
        assert!(docs.join(".gitattributes").exists());
        assert!(docs.join("biome.json").exists());
        // biome.json byte-exact.
        assert_eq!(
            std::fs::read_to_string(docs.join("biome.json")).unwrap(),
            "{\n  \"$schema\": \"https://biomejs.dev/schemas/2.0.0/schema.json\",\n  \"files\": {\n    \"ignore\": [\n      \"**/*\"\n    ]\n  }\n}\n"
        );
        // .gitignore patched with a marker block; .prettierignore untouched.
        let gi = std::fs::read_to_string(dir.path().join(".gitignore")).unwrap();
        assert!(gi.contains("# ask:start"));
        assert!(gi.contains(".ask/docs/"));
        assert!(!dir.path().join(".prettierignore").exists());
    }

    #[test]
    fn remove_strips_nested_and_root() {
        let dir = tempfile::tempdir().unwrap();
        opt_in(dir.path());
        std::fs::write(dir.path().join(".gitignore"), "node_modules\n").unwrap();
        manage_ignore_files(dir.path(), IgnoreMode::Install).unwrap();
        manage_ignore_files(dir.path(), IgnoreMode::Remove).unwrap();

        assert!(!get_docs_dir(dir.path()).join("biome.json").exists());
        let gi = std::fs::read_to_string(dir.path().join(".gitignore")).unwrap();
        assert!(!gi.contains("# ask:start"));
        assert!(gi.contains("node_modules"));
    }

    #[test]
    fn nested_write_is_idempotent() {
        let dir = tempfile::tempdir().unwrap();
        opt_in(dir.path());
        assert!(!write_nested_configs(dir.path()).unwrap().is_empty());
        // Second run: content unchanged → nothing reported.
        assert!(write_nested_configs(dir.path()).unwrap().is_empty());
    }
}
